src/Verismith/Verilog2005/AST.hs

```diff
@@ -0,0 +1,1583 @@
+  | NTWire
+  | NTWAnd
+  | NTWOr
+  deriving (Eq, Ord, Bounded, Enum, Data, Generic)
+
+instance Show NetType where
+  show x = case x of
+    NTSupply1 -> "supply1"
+    NTSupply0 -> "supply0"
+    NTTri -> "tri"
+    NTTriAnd -> "triand"
+    NTTriOr -> "trior"
+    NTTri1 -> "tri1"
+    NTTri0 -> "tri0"
+    NTUwire -> "uwire"
+    NTWire -> "wire"
+    NTWAnd -> "wand"
+    NTWOr -> "wor"
+
+-- | Net drive strengths
+data Strength = StrSupply | StrStrong {-default-} | StrPull | StrWeak
+  deriving (Eq, Ord, Data, Generic)
+
+instance Show Strength where
+  show x = case x of
+    StrSupply -> "supply"
+    StrStrong -> "strong"
+    StrPull -> "pull"
+    StrWeak -> "weak"
+
+data DriveStrength
+  = DSNormal
+      { _DS0 :: !Strength,
+        _DS1 :: !Strength
+      }
+  | DSHighZ
+      { _DSHZ :: !Bool,
+        _DSStr :: !Strength
+      }
+  deriving (Show, Eq, Ord, Data, Generic)
+
+dsDefault = DSNormal {_DS0 = StrStrong, _DS1 = StrStrong}
+
+-- | Capacitor charge
+data ChargeStrength = CSSmall | CSMedium {-default-} | CSLarge
+  deriving (Eq, Ord, Bounded, Enum, Data, Generic)
+
+instance Show ChargeStrength where
+  show x = case x of CSSmall -> "(small)"; CSMedium -> "(medium)"; CSLarge -> "(large)"
+
+-- | Left side of assignments
+data LValue dr
+  = LVSingle
+      { _LVIdent :: !HierIdent,
+        _LVDimRange :: !(Maybe dr)
+      }
+  | LVConcat !(NonEmpty (LValue dr))
+  deriving (Show, Eq, Ord, Data, Generic)
+
+type NetLValue = LValue CDimRange
+
+type VarLValue = LValue DimRange
+
+-- $(makePlates ''LValue)
+
+-- | Net assignment
+data NetAssign = NetAssign {_NALValue :: !NetLValue, _NAValue :: !Expr}
+  deriving (Show, Eq, Ord, Data, Generic)
+
+-- | Variable assignment
+data VarAssign = VarAssign {_VALValue :: !VarLValue, _VAValue :: !Expr}
+  deriving (Show, Eq, Ord, Data, Generic)
+
+-- | Parameter assignment list
+data ParamAssign
+  = ParamPositional ![Expr]
+  | ParamNamed ![Identified (Maybe MinTypMax)]
+  deriving (Show, Eq, Ord, Data, Generic)
+
+-- | Port assignment list
+data PortAssign
+  = PortNamed ![AttrIded (Maybe Expr)]
+  | PortPositional ![Attributed (Maybe Expr)]
+  deriving (Show, Eq, Ord, Data, Generic)
+
+-- | Event primitive
+data EventPrim = EventPrim {_EPOp :: !EventPrefix, _EPExpr :: !Expr}
+  deriving (Show, Eq, Ord, Data, Generic)
+
+-- | Event control
+data EventControl
+  = ECIdent !HierIdent
+  | ECExpr !(NonEmpty EventPrim)
+  | ECDeps
+  deriving (Show, Eq, Ord, Data, Generic)
+
+-- | Delay or Event control
+data DelayEventControl
+  = DECDelay !Delay1
+  | DECEvent !EventControl
+  | DECRepeat
+      { _DECRExpr :: !Expr,
+        _DECREvent :: !EventControl
+      }
+  deriving (Show, Eq, Ord, Data, Generic)
+
+-- | Procedural continuous assignment
+data ProcContAssign
+  = PCAAssign !VarAssign
+  | PCADeassign !VarLValue
+  | PCAForce !(Either VarAssign NetAssign)
+  | PCARelease !(Either VarLValue NetLValue)
+  deriving (Show, Eq, Ord, Data, Generic)
+
+-- $(makePlates ''NetAssign)
```

Comment:
Were there any problems with adding the `Plates` typeclass?

Reply:
I did not try adding Plates because I don't understand how to use it effectively. For instance I don't know how to use it as a mean to retrieve the identifiers of specific elements (all wires names but not function names or the opposite, only module names).
I only put it commented at the places I think it will likely be used. If you have more insights you can modify it how you want.

Reply:
That sounds good, and yeah it needs to be type driven, so function identifiers should have a different type to wire identifiers for plates to work nicely, but we can look at that in another PR at some point, and I can give it a go adding them.